### PR TITLE
cNFT Metadata

### PR DESCRIPTION
**Pull request by devey**
Try different approaches to storing `Metadata`, taking inspiration from `near-sdk-as`.
Create a structure for `TokenMetadata`. Create structure and interface method for `ContractMetadata`.

### CODE REVIEW

File: as/cNFT/assembly/models/persistent_nft_contract_metadata.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        reference_hash: '',
+        packages_script: '',
+        render_script: '',
+        style_css: '',
+        parameters: '',
+    }
+}
+
+@nearBindgen
+export class PersistentNFTContractMetadata {
```
Comment (achrafismyname) [inline as/cNFT/assembly/models/persistent_nft_contract_metadata.ts:39]: Why not call this `PersistentContractMetadata` like `PersistentTokenMetadata`?
The same for `NFTContractMetadata` and the file name

File: as/cNFT/assembly/models/persistent_tokens_metadata.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    starts_at: string
+    updated_at: string
+    reference: string
+    reference_hash: string
+}
+
+export function defaultTokenMetadata(): TokenMetadata {
+    return {
+        title: '',
+        issued_at: '',
```
Comment (achrafismyname) [inline as/cNFT/assembly/models/persistent_tokens_metadata.ts:23]: Having empty strings as default doesn't seem right, why can't it be `null`?

File: as/cNFT/assembly/models/persistent_tokens.ts [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
+import { PersistentSet, PersistentUnorderedMap, u128 } from 'near-sdk-as'
 import { AccountId, TokenId } from '../types'
-import { Token } from './token'
+
+@nearBindgen
+export class Token {
+    id: string
+    owner_id: string
+    creator: string
+    prev_owner: string
```
Comment (achrafismyname) [inline as/cNFT/assembly/models/persistent_tokens.ts:9]: Shouldn't `creator` be called `creator_id` and `prev_owner` called `prev_owner`. 
To keep it the consistent with `owner_id`

File: as/cNFT/assembly/mint.ts [hunk 4]
```diff
@@ -1,30 +1,32 @@
...
     //     assert_deposit_attached(u128.from(0))
 
-    /** TODO Assert uniqueId is actually unique */
-
     let token = new Token()
 
-    token.media = tokenMetadata.media
-    token.extra = tokenMetadata.extra
-    /** TODO not always */
+    /**@todo Not always sender is creator i guess, would rather pass it to function */
```
Comment (achrafismyname) [inline as/cNFT/assembly/mint.ts:14]: If we pass the creator to mint, won't this make it easy to fake tokens? 
Someone can mint a token using the name of a popular creator

**Review by achrafismyname — COMMENTED** [review]
Seems good, I just have some questions to make sure

File: as/cNFT/assembly/index.ts [hunk 5]
```diff
@@ -1,13 +1,28 @@
-import { persistent_tokens } from './models/persistent_tokens'
+import { storage } from 'near-sdk-as'
```
Comment (AshenI99) [inline as/cNFT/assembly/index.ts:1]: As I see `storage` is not used anywhere inside the file. So is it necessary to include here?

File: as/cNFT/assembly/mint.ts [hunk 6]
```diff
@@ -1,30 +1,32 @@
...
     token.creator = context.sender
```
Comment (AshenI99) [inline as/cNFT/assembly/mint.ts:15]: Here we are filling only the `creator` in token. But token has several properties like `id`, `owner_id` and `prev_owner`. So we are returning token without those properties. Is it okay ?

**Review by AshenI99 — COMMENTED** [review]
Other than the couple of things I've mentioned, I would agree with @achrafismyname regarding `null` values instead of `empty` strings. So It is better to use `null` values if it is not affecting to something. Other things seems to be fine. That's all as for now. If I suspect anything more, I'll let you know.

File: as/cNFT/assembly/models/persistent_nft_contract_metadata.ts (hunk 1 above)
Comment (devey) [reply as/cNFT/assembly/models/persistent_nft_contract_metadata.ts:39]: This is naming used by `NEAR` team: https://nomicon.io/Standards/NonFungibleToken/Metadata.html
Agree it's not perfect tho!

File: as/cNFT/assembly/models/persistent_tokens_metadata.ts (hunk 2 above)
Comment (devey) [reply as/cNFT/assembly/models/persistent_tokens_metadata.ts:23]: Not sure why but this doesn't work (tried a long time ago too):
```
ERROR TS2322: Type '~lib/string/String | null' is not assignable to type '~lib/string/String'.
```

Could you open an issue for it actually? Time to figure out why it's not working and write it down somewhere or fix it.

File: as/cNFT/assembly/models/persistent_tokens.ts (hunk 3 above)
Comment (devey) [reply as/cNFT/assembly/models/persistent_tokens.ts:9]: Yes, better :)

File: as/cNFT/assembly/index.ts (hunk 5 above)
Comment (devey) [reply as/cNFT/assembly/index.ts:1]: :) cleaned

File: as/cNFT/assembly/models/persistent_nft_contract_metadata.ts (hunk 1 above)
Comment (devey) [reply as/cNFT/assembly/models/persistent_nft_contract_metadata.ts:39]: @achrafismyname @AshenI99

File: as/cNFT/assembly/models/persistent_tokens_metadata.ts (hunk 2 above)
Comment (devey) [reply as/cNFT/assembly/models/persistent_tokens_metadata.ts:23]: @achrafismyname @AshenI99

File: as/cNFT/assembly/mint.ts (hunk 4 above)
Comment (devey) [reply as/cNFT/assembly/mint.ts]: Updated the comment. This is just a skeleton, left the comment as a reminder. You can see [here](https://github.com/near-examples/nft-tutorial/blob/4093906440bf8688d1762e8b988688fcd5be5145/nft-contract/src/mint.rs#L6) on approach to minting. Feel free to research others and discuss them on GitHub.

File: as/cNFT/assembly/models/persistent_tokens_metadata.ts (hunk 2 above)
Comment (AshenI99) [reply as/cNFT/assembly/models/persistent_tokens_metadata.ts:23]: Done :-)

**Pull request merged**